src/windows/systemw.c

```diff
@@ -230,6 +230,23 @@ void MTY_SetCrashFunc(MTY_CrashFunc func, void *opaque)
 	SYSTEM_OPAQUE = opaque;
 }
 
+static BOOL WINAPI system_handler_routine(DWORD dwCtrlType)
+{
+	// Console close == SIGTERM, Ctrl+C == SIGINT
+	if (SYSTEM_CRASH_FUNC)
+		SYSTEM_CRASH_FUNC(dwCtrlType == CTRL_CLOSE_EVENT || dwCtrlType == CTRL_C_EVENT, SYSTEM_OPAQUE);
+
+	// Prevent secondary calls to the SYSTEM_CRASH_FUNC
+	SYSTEM_CRASH_FUNC = NULL;
```

Comment:
I don't know if you need to protect against multi-threaded scenarios, but perhaps you should do an atomic exchange of the crash func pointer into a local variable while setting the static variable to NULL.